From 741b1363f4dedf84a53546d46494ad0da3524579 Mon Sep 17 00:00:00 2001
From: YumeMichi <do4suki@gmail.com>
Date: Mon, 23 Mar 2020 15:30:06 +0800
Subject: [PATCH] bionic: Allow disable memcpy opt for certain targets

 * Based on "libc: memcpy_base: Disable opt for certain targets"

Signed-off-by: YumeMichi <do4suki@gmail.com>
Change-Id: I798af182809e619ce1b02491c442938c2190b254
---
 libc/Android.bp                             | 10 ++++++++++
 libc/arch-arm/dynamic_function_dispatch.cpp |  4 ++++
 2 files changed, 14 insertions(+)

diff --git a/libc/Android.bp b/libc/Android.bp
index 6031b1661..0783715f2 100644
--- a/libc/Android.bp
+++ b/libc/Android.bp
@@ -1439,6 +1439,16 @@ cc_library_static {
         },
     },
 
+    product_variables: {
+        lineage: {
+            cpu_memcpy_base_opt_disable: {
+                cflags: [
+                    "-DCPU_MEMCPY_BASE_OPT_DISABLE",
+                ],
+            },
+        },
+    },
+
     whole_static_libs: [
         "libc_common",
     ],
diff --git a/libc/arch-arm/dynamic_function_dispatch.cpp b/libc/arch-arm/dynamic_function_dispatch.cpp
index 640f3306a..0278b6b4f 100644
--- a/libc/arch-arm/dynamic_function_dispatch.cpp
+++ b/libc/arch-arm/dynamic_function_dispatch.cpp
@@ -172,7 +172,11 @@ DEFINE_IFUNC(__memcpy) {
         case kCortexA9:
             RETURN_FUNC(__memcpy_func, __memcpy_a9);
         case kKrait:
+#ifdef CPU_MEMCPY_BASE_OPT_DISABLE
+            RETURN_FUNC(__memcpy_func, __memcpy_a15);
+#else
             RETURN_FUNC(__memcpy_func, __memcpy_krait);
+#endif
         case kCortexA53:
             RETURN_FUNC(__memcpy_func, __memcpy_a53);
         case kCortexA55:
-- 
2.25.1

